[eas-cli] add channel:delete

jkhales opened the pull request:
<!-- Thanks for contributing to _EAS CLI_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

Add eas channel:delete

# How

copied eas branch:delete

# Test Plan

tested in demo repo

jkhales commented:
/changelog-entry new-feature Add eas channel:create

jkhales commented on packages/eas-cli/src/commands/channel/delete.ts:
```diff
@@ -0,0 +1,148 @@
+    let {
+      args: { name },
+      flags: { json: jsonFlag },
+    } = this.parse(ChannelDelete);
```
Ah I see, OK. Going to put us a another PR to correct this in a bunch of other commands as well.

jkhales replied:
https://github.com/expo/eas-cli/pull/848

dsokal commented on packages/eas-cli/src/commands/channel/delete.ts:
```diff
@@ -0,0 +1,152 @@
+  ];
+  static flags = {
+    json: flags.boolean({
+      description: `return JSON with the edited branch's ID and name.`,
```
edited?

dsokal commented on packages/eas-cli/src/commands/channel/delete.ts:
```diff
@@ -0,0 +1,152 @@
+  };
+
+  async runAsync(): Promise<void> {
+    let {
```
```suggestion
    const {
```
Don't use `let` if you don't need it.

dsokal commented on packages/eas-cli/src/commands/channel/delete.ts:
```diff
@@ -0,0 +1,152 @@
+      if (jsonFlag) {
+        throw new Error(validationMessage);
+      }
+      ({ name } = await promptAsync({
```
Do not reassign variables. IMO it's an anti-pattern and makes refactoring a nightmare.

dsokal commented on packages/eas-cli/src/commands/channel/delete.ts:
```diff
@@ -0,0 +1,152 @@
+
+    if (!name) {
+      const validationMessage = 'Channel name may not be empty.';
+      if (jsonFlag) {
```
Hmmm, I'm not sure if we want to throw an error if someone passes the `--json` flag. I'd prefer to add another flag - `--non-interactive` - and throw only then.

dsokal commented on packages/eas-cli/src/commands/channel/delete.ts:
```diff
@@ -0,0 +1,152 @@
+      throw new Error(`Could not find channel ${name} on ${fullName}`);
+    }
+
+    if (!jsonFlag) {
```
Same here, this looks weird. I'm 100% sure the logic shouldn't be based on the `--json` flag.
We should either don't let users run this in non-interactive mode or handle the `--non-interactive` flag. Is there a use case for running this on CI?

jkhales replied:
switched to `non-interactive`

dsokal commented on packages/eas-cli/src/commands/channel/delete.ts:
```diff
@@ -0,0 +1,152 @@
+      Log.newLine();
+      const confirmed = await toggleConfirmAsync({ message: 'Are you sure you wish to proceed?' });
+      if (!confirmed) {
+        Log.error(`Cancelled deletion of channel: "${name}".`);
```
```suggestion
        Log.error(`Canceled deletion of channel: "${name}".`);
```
🇺🇸

dsokal commented on packages/eas-cli/src/commands/channel/delete.ts:
```diff
@@ -0,0 +1,152 @@
+      printJsonOnlyOutput(deletionResult);
+    }
+
+    Log.withTick(`️Deleted channel "${name}".`);
```
Do we want to print this with `--json` flag?

jkhales replied:
moved behind an if/else

jkhales commented on packages/eas-cli/src/commands/channel/delete.ts:
```diff
@@ -0,0 +1,152 @@
+    graphqlClient
+      .query<GetChannelInfoQuery, GetChannelInfoQueryVariables>(
+        gql`
+          query GetChannelInfo($appId: String!, $name: String!) {
```
There is an extra database query on the most similar one I could find:
<img width="604" alt="Screen Shot 2021-12-16 at 3 39 08 PM" src="https://user-images.githubusercontent.com/1220444/146464806-1c2c15b0-d1a1-41e5-a042-978b62de97c7.png">

But I guess this is an infrequent command and it wouldn't be so bad to combine these commands in the name of deduplication... on the other hand the future of these two functions are not inherently tied so there isn't a great reason to unify them.

Do you feel strongly about combining them?

dsokal reviewed (approved):
I don't like the behavior change with `--json` flag but 🤷

jkhales commented:
> I don't like the behavior change with `--json` flag but 🤷

I think you are right.

I didn't get around to addressing those comments yesterday, sorry if it came off as me ignoring them!

I'm going to push some changes when I get back from lunch and would greatly appreciate you taking a look at them and also making sure I am using the correct english 🇺🇸 😄

jkhales commented on packages/eas-cli/src/commands/channel/delete.ts:
```diff
@@ -0,0 +1,164 @@
+      args: { name: nameArg },
+      flags: { json: jsonFlag, 'non-interactive': nonInteractiveFlag },
+    } = this.parse(ChannelDelete);
+    if (jsonFlag && !nonInteractiveFlag) {
+      error('--json is allowed only in non-interactive mode', { exit: 1 });
+    }
```
There is an argument that setting `--json` should automatically set `nonInteractiveFlag = true`, but for now am following the convention set here: https://github.com/expo/eas-cli/blob/main/packages/eas-cli/src/commands/build/index.ts#L107

dsokal reviewed (approved):
👏
